### ISSUE TITLE
CoverageScript.command_line wipes out 1st sys.path entry.

### ISSUE THREAD

**Issue by jsirois** [coveragepy/coveragepy#715]
This can be problematic if the coverage distribution itself is not the first item on the classpath:
https://github.com/nedbat/coveragepy/blob/401471c1c51c2a99b31b0508578d04b4da03eebb/coverage/cmdline.py#L468-L470

Is there any reason this couldn't be a `sys.path.insert(0, '')`?

I'm happy to do the work but I just wanted to run this by you first.

**Comment by nedbat** [comment]
I'd feel better about changing this if you could show me a bad behavior that the current code causes.

**Comment by jsirois** [comment]
OK.

CI is here, but its a wall of text: https://travis-ci.org/pantsbuild/pants/jobs/437335398

The relevant bug presentation is configuring coverage with a plugin, that plugin being on the sys.path (in the 1st slot as it turns out) and then coverage mysteriously not being able to find the plugin::
```
11:20:19 00:02       [coverage-html]
                     pex: PEX.run invoking /home/jsirois/dev/pantsbuild/jsirois-pants/build-support/pants_dev_deps.venv/bin/python2.7 /home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/bdd83c141bb9cfffde291c49aa9e41781cc24326 html -i --rcfile /home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest/testprojects.tests.python.pants.constants_only.constants_only/tmpW10vvZ -d /home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest/testprojects.tests.python.pants.constants_only.constants_only/coverage
Traceback (most recent call last):
                       File "/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/bdd83c141bb9cfffde291c49aa9e41781cc24326/.bootstrap/_pex/pex.py", line 351, in execute
                         self._wrap_coverage(self._wrap_profiling, self._execute)
                       File "/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/bdd83c141bb9cfffde291c49aa9e41781cc24326/.bootstrap/_pex/pex.py", line 281, in _wrap_coverage
                         runner(*args)
                       File "/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/bdd83c141bb9cfffde291c49aa9e41781cc24326/.bootstrap/_pex/pex.py", line 313, in _wrap_profiling
                         runner(*args)
                       File "/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/bdd83c141bb9cfffde291c49aa9e41781cc24326/.bootstrap/_pex/pex.py", line 390, in _execute
                         return self.execute_entry(self._pex_info_overrides.entry_point)
                       File "/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/bdd83c141bb9cfffde291c49aa9e41781cc24326/.bootstrap/_pex/pex.py", line 514, in execute_entry
                         return runner(entry_point)
                       File "/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/bdd83c141bb9cfffde291c49aa9e41781cc24326/.bootstrap/_pex/pex.py", line 532, in execute_pkg_resources
                         return runner()
                       File "/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/coverage-4.5.1-cp27-cp27mu-manylinux1_x86_64.whl/coverage/cmdline.py", line 755, in main
                         status = CoverageScript().command_line(argv)
                       File "/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/coverage-4.5.1-cp27-cp27mu-manylinux1_x86_64.whl/coverage/cmdline.py", line 507, in command_line
                         self.coverage.load()
                       File "/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/coverage-4.5.1-cp27-cp27mu-manylinux1_x86_64.whl/coverage/control.py", line 676, in load
                         self._init()
                       File "/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/coverage-4.5.1-cp27-cp27mu-manylinux1_x86_64.whl/coverage/control.py", line 227, in _init
                         self.plugins = Plugins.load_plugins(self.config.plugins, self.config, self.debug)
                       File "/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/coverage-4.5.1-cp27-cp27mu-manylinux1_x86_64.whl/coverage/plugin_support.py", line 41, in load_plugins
                         __import__(module)
                     ImportError: No module named __pants_backend_python_tasks_pytest_prep__
```

At the handoff point from executing code, the particular sys.path is:
```
 >>> Executing coverage.cmdline:main using sys.path:
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/bdd83c141bb9cfffde291c49aa9e41781cc24326
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/build-support/pants_dev_deps.venv/lib/python27.zip
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/build-support/pants_dev_deps.venv/lib/python2.7
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/build-support/pants_dev_deps.venv/lib/python2.7/plat-linux2
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/build-support/pants_dev_deps.venv/lib/python2.7/lib-tk
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/build-support/pants_dev_deps.venv/lib/python2.7/lib-old
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/build-support/pants_dev_deps.venv/lib/python2.7/lib-dynload
                     	/usr/lib64/python2.7
                     	/usr/lib/python2.7
                     	/usr/lib/python2.7/plat-linux2
                     	/usr/lib64/python2.7/lib-tk
                     	/usr/lib/python2.7/lib-tk
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/pyprep/requirements/CPython-2.7.15/5e0b9a73f7766ef6381bac7241ece3a52aefebf3-DefaultFingerprintStrategy_f4bc9a770d1b
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/pyprep/sources/e9d7f65130ec5c9a2a5d05db01a0075427938983
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/pytest-3.6.4-py2.py3-none-any.whl
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/setuptools-40.4.3-py2.py3-none-any.whl
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/funcsigs-1.0.2-py2.py3-none-any.whl
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/linecache2-1.0.0-py2.py3-none-any.whl
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/pytest_timeout-1.2.1-py2.py3-none-any.whl
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/attrs-18.2.0-py2.py3-none-any.whl
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/pytest_cov-2.4.0-py2.py3-none-any.whl
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/more_itertools-4.3.0-py2-none-any.whl
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/unittest2-1.1.0-py2.py3-none-any.whl
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/six-1.11.0-py2.py3-none-any.whl
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/atomicwrites-1.2.1-py2.py3-none-any.whl
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/py-1.6.0-py2.py3-none-any.whl
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/coverage-4.5.1-cp27-cp27mu-manylinux1_x86_64.whl
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/pluggy-0.7.1-py2.py3-none-any.whl
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/argparse-1.4.0-py2.py3-none-any.whl
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/traceback2-1.4.0-py2.py3-none-any.whl
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/pyprep/requirements/CPython-2.7.15/5e0b9a73f7766ef6381bac7241ece3a52aefebf3-DefaultFingerprintStrategy_f4bc9a770d1b/.deps/six-1.11.0-py2.py3-none-any.whl
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/pyprep/requirements/CPython-2.7.15/5e0b9a73f7766ef6381bac7241ece3a52aefebf3-DefaultFingerprintStrategy_f4bc9a770d1b/.deps/thrift-0.11.0-cp27-cp27mu-linux_x86_64.whl
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/bdd83c141bb9cfffde291c49aa9e41781cc24326/.bootstrap
```

Here the 1st entry, `/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/bdd83c141bb9cfffde291c49aa9e41781cc24326` is the executing code and it contains the plugin.
Coverage is on the sys.path in the `/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/coverage-4.5.1-cp27-cp27mu-manylinux1_x86_64.whl` entry lower down.

I went down the rabbit hole and things were ok at coverage/cmdline.py line 755:
```
>>> In coverage.cmdline with sys.path:
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/bdd83c141bb9cfffde291c49aa9e41781cc24326
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/build-support/pants_dev_deps.venv/lib/python27.zip
...
```

But not ok here:
```
>>> In coverage.cmdline about to call self.coverage.load() with sys.path:
                     	
                     	/home/jsirois/dev/pantsbuild/jsirois-pants/build-support/pants_dev_deps.venv/lib/python27.zip
...
```

**Comment by jsirois** [comment]
FWIW - I'm hacking around presently by ensuring the 1st sys.path entry (the user code), is on the sys.path twice.

**Comment by jsirois** [comment]
Also probably a wall of text, but the fix commit downstream is here: https://github.com/pantsbuild/pants/pull/6594/commits/2d957560ff9f977561e4b5f6ca15d1a2a0112296

**Comment by nedbat** [comment]
The problem with making the change you suggest is that then "coverage run prog.py" creates a different sys.path than "python prog.py".  Making the two the same has been my guiding star, since it is difficult to decide what the "correct" behavior should be other than that.

I don't know what to do about your plugin problem....

**Comment by jsirois** [comment]
I do have a workaround - so, bottom line, no worries. That said, commenting out the `sys.path[0]` mutation leads to the same sys.path via `python test.py` or `coverage run test.py` since python places CWD at the head of `sys.path` already. IOW `sys.path[0] = ''` is a no-op in this vanilla case. Presumably it fixes up a `sys.path` in a less standard case. The comment references django which I have little familiarity with. It would seem though that there is an assumption that whatever is in the 0th slot is unimportant. I'm not sure how coverage can be confident what is in that slot (except in the vanilla case mentioned above, in which case the line is a no-op). Perhaps the django case puts a safely-nukeable path element there?

**Comment by nedbat** [comment]
@jsirois Hmm, commenting out that line might be fine: the coverage.py tests all pass.  Does it solve your problem?  There is another sys.path manipulation later when we exec the file.

**Comment by jsirois** [comment]
I'm not positive, but I don't think it will affects us. Here I assume?:
https://github.com/nedbat/coveragepy/blob/7a099a2204e6863824f11432c807442f3ad5306c/coverage/execfile.py#L174-L176

We use pytest-cov to actually collect coverage data under the pytest runner and they appear to use the coverage API (ie: coverage.run(), etc) - not execfile. We only have this issue trying to produce reports after the coverage run using the coverage command reporting tools.

**Comment by nedbat** [comment]
@jsirois Could you try removing the line?

**Comment by jsirois** [comment]
Will do. If I don't get a PR out tonight I will tomorrow. Thanks Ned!

**Comment by nedbat** [comment]
A PR isn't necessary (the change is simple enough), but I'd like to know how your scenario behaves with the change.

**Comment by jsirois** [comment]
Ah, gotcha. I can guaranty it solves our problem. The digging I hinted at above with the injected print statements proved that. That said, I can back up that guaranty with a full test by tomorrow. I'll report back.

**Comment by nedbat** [comment]
@jsirois #678 is also about this.

**Comment by jsirois** [comment]
Aha, yes I duped #678.

I just had a chance to burn a local coverage dist with:
```diff
$ git diff coverage/cmdline.py
diff --git a/coverage/cmdline.py b/coverage/cmdline.py
index edbc1d25..e6ea6e23 100644
--- a/coverage/cmdline.py
+++ b/coverage/cmdline.py
@@ -465,10 +465,6 @@ class CoverageScript(object):
         if self.do_help(options, args, parser):
             return OK
 
-        # We need to be able to import from the current directory, because
-        # plugins may try to, for example, to read Django settings.
-        sys.path[0] = ''
-
         # Listify the list options.
         source = unshell_list(options.source)
         omit = unshell_list(options.omit)
```

That did in fact solve the problem pants was having and coverage now works without hacks.

**Comment by nedbat** [comment]
Fixed in https://github.com/nedbat/coveragepy/commit/e82d24da2355df86fc46d7d41142ce308ed2c919.

**Comment by jsirois** [comment]
Thanks @nedbat!

**Comment by nedbat** [comment]
Hmm, wait a minute: looks like this breaks Windows (and my Windows CI is missing some builds!?).  Re-opening for the moment.

**Comment by nedbat** [comment]
It wasn't a Windows problem.  Coverage.py tests fail if you don't run them with pytest-xdist, reported here: https://github.com/pytest-dev/pytest-xdist/issues/376.  I'm still looking for the right solution.

**Comment by nedbat** [comment]
@jsirois Looking into this more, I have a few questions.  The sys.path docs (https://docs.python.org/3/library/sys.html#sys.path) say, "As initialized upon program startup, the first item of this list, path[0], is the directory containing the script that was used to invoke the Python interpreter."  What is sys.path[0] for you, and why is it different than this?

Also, I am just curious about your coverage plugin: what does it do?

**Comment by jsirois** [comment]
@nedbat we run pytest and coverage from within a pex - roughly a zipapp, and so the sys.path[0] entry points to our pex.

In other words - in the [debug output above](https://github.com/nedbat/coveragepy/issues/715#issuecomment-427415660) snipped and annotated here:
```
[0] /home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/bdd83c141bb9cfffde291c49aa9e41781cc24326
/usr/lib64/python2.7
...
/usr/lib/python2.7/lib-tk
[12] /home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4
[13] /home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/pyprep/requirements/CPython-2.7.15/5e0b9a73f7766ef6381bac7241ece3a52aefebf3-DefaultFingerprintStrategy_f4bc9a770d1b
[14] /home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/pyprep/sources/e9d7f65130ec5c9a2a5d05db01a0075427938983
/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/pytest-3.6.4-py2.py3-none-any.whl
...
/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/py-1.6.0-py2.py3-none-any.whl
[27] /home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/coverage-4.5.1-cp27-cp27mu-manylinux1_x86_64.whl
/home/jsirois/dev/pantsbuild/jsirois-pants/.pants.d/test/pytest-prep/CPython-2.7.15/15d60b3366ef43c7a4ae6b5f5e2d7567a95fe5f4/.deps/pluggy-0.7.1-py2.py3-none-any.whl
...
```
The 0th entry is our pex, then a bunch of system python entries, then more bits of our pex (we combine 4) in 12, 13 & 14 - then 3rdparty dependencies of the pex - including coverage down in entry 27. The critical point which applies here afaict is coverage is not guaranteed to be in the 0th slot if some other tool is running coverage. As such coverage should not blow away any sys.path entry it is not sure it owns. In this case the 0th entry is us, the bootstrap code that runs coverage and contains the coverage plugins we wish to activate. More generally, inserting or appending a sys.path entry as is appropriate to the case at hand is almost always what you want to do afaict. If there might be multiple versions of the thing and you need to have your version seen, insert 0, if you know you're unique and just need to be visible, append.

The plugin just attempts to remap filenames. The scenario is as follows:
1. A user wants to run tests and get coverage data for those tests.
2. Pants (a build tool), gathers up the relevant test code and production code under test and forms a hermetic pex containing the user code as well as pytest and coverage.
3. Pants executes pytest and coverage against the user code in the context of the pex.

Because the coverage is gathered in step 3, it sees and reports on paths different from the original loose filesystem paths the user expects; and as such we have a coverage plugin that remaps pex pathnames back to user-space.

**Comment by nedbat** [comment]
@jsirois Thanks for the explanation.

BTW, your description of your plugin sounds like what the `[paths]` configuration setting does: https://coverage.readthedocs.io/en/v4.5.x/config.html#paths .   Or is your situation more complex?

**Comment by jsirois** [comment]
The situation is a bit more complex. See a comment here: https://github.com/pantsbuild/pants/issues/5426#issuecomment-363593177 and our discussion in #646.

The issue boils down to this code here:
https://github.com/nedbat/coveragepy/blob/e6b883903e6f0eb396d35d43ff82c3c5644af84c/coverage/control.py#L559-L565

Pants runs tests for projects that can be split over multiple source trees; ie:
```
project/
  widget/src/python/
    widget/__init__.py  # A namespace package
    widget/module1.py
  util/src/python
    widget/__init__.py  # A namespace package
    widget/util.py
```
As such, a test against the `widget` package in this example can produce coverage that is spanned multiple different local source roots. The PathAliases code does not allow for this as things stand and 1 local source root "wins" leading to trampling of data from the losers. So my options were to contribute a patch to support the concept of multiple local source roots or else write a plugin. I'd be happy to ditch the plugin in favor of using `combine` if you think a patch to add support for this concept is desirable.

**Comment by nedbat** [comment]
This is re-fixed in commit b7e0eec3d794ee350f8a4392a0c9cc90fff54fa0, released as part of 5.0a4.
